opencog/atomspace/Handle.h

```diff
@@ -71,13 +73,33 @@ friend class AtomspaceHTabler;    // persistance
     AtomPtr resolve_ptr();
     static void set_resolver(const AtomTable*);
     static void clear_resolver(const AtomTable*);
+
+    // Set the UUID in a handle. This is the safe version. Use  this version
+    // if you have any doubts at all about the possibility that the handle
+    // may have been resolved. This sets the new UUID and also calls 
+    // resets the shared_ptr AtomPtr so it doesn't point to anything,
+    // the reference is decremented, and the object deleted if this is the
+    // last one. This is used by AtomTable to avoid having to construct
+    // temporaries while iterating when the iteration itself can resolve the
+    // AtomPtr in order to clear the AtomPtr reference so it won't be pointing
+    // to an atom that has a UUID that is different than the UUID passed
+    // to set_uuid.    
+    void set_uuid(const UUID uuid) { _ptr.reset(); _uuid = uuid; }
```

Comment:
Note that `TypeIndex` and `TargetTypeIndex` subclass from `FixedIntegerIndex` and the internal index for `FixedIntegerIndex` is:

```
std::vector<UnorderedUUIDSet> idx; 
```
from: https://github.com/opencog/opencog/blob/master/opencog/atomspace/FixedIntegerIndex.h#L44

So we've got a set of `UUID`s at the core of the `TypeIndex` and `TargetTypeIndex`. This is turned into a handle here in `TypeIndex::iterator::operator*` which would be quite easy to miss during a checkin:

https://github.com/opencog/opencog/blob/master/opencog/atomspace/TypeIndex.cc#L106

and since TargetTypeIndex subclasses TypeIndex:

https://github.com/opencog/opencog/blob/master/opencog/atomspace/TargetTypeIndex.h#L42

you end up with the same operator *. So it looks like these use Handles, but they don't.